sidecar/codec.go

```diff
@@ -15,77 +16,107 @@ const (
 )
 
 var (
-	// defaultStringEncoding is the default encoding algorithm we use to
-	// encode the gzipped binary data as a string.
-	defaultStringEncoding = base64.URLEncoding
-
-	// sidecarPrefix is a specially prepared prefix that will spell out
-	// "sidecar" when encoded as base64. It will cause all our sidecar
-	// tickets to be easily recognizable with a human readable part.
-	sidecarPrefix = []byte{0xb2, 0x27, 0x5e, 0x71, 0xaa, 0xc0}
+	// sidecarPrefix will serve as our "human readable" prefix to make the
+	// strings easily identifiable.
+	sidecarPrefix = "sidecar"
+
+	// encodingVersion is the current encoding version used so each ticket
+	// will have a predictable character after the sidecar prefix.
+	encodingVersion = []byte{0}
 )
 
 // EncodeToString serializes and encodes the ticket as an URL safe string that
 // contains a human readable prefix and a checksum.
 func EncodeToString(t *Ticket) (string, error) {
-	var buf bytes.Buffer
-
-	// Add a prefix to make the ticket easily recognizable by human eyes.
-	if _, err := buf.Write(sidecarPrefix); err != nil {
+	var (
+		checksumBuf bytes.Buffer
+		encodeBuf   bytes.Buffer
+	)
+
+	// First, we'll write the sidecar prefix, as well as the serialized
+	// ticket into the buffer that we'll use to generate the checksum. We
+	// do this as we won't encode the checksum using base58.
+	if _, err := checksumBuf.Write([]byte(sidecarPrefix)); err != nil {
 		return "", err
 	}
-
-	// Serialize the raw ticket itself.
-	if err := SerializeTicket(&buf, t); err != nil {
+	if _, err := checksumBuf.Write(encodingVersion); err != nil {
+		return "", err
+	}
+	if err := SerializeTicket(&checksumBuf, t); err != nil {
 		return "", err
 	}
 
-	// Create a checksum (SHA256) of all the bytes so far (prefix + raw
-	// ticket) and add 4 bytes of that to the string encoded version. This
-	// will prevent the ticket from appearing valid even if parts of it were
-	// not copy/pasted correctly.
-	hash := sha256.New()
-	_, _ = hash.Write(buf.Bytes())
-	checksum := hash.Sum(nil)[0:4]
-	if _, err := buf.Write(checksum); err != nil {
+	// We'll now hash our full checksum buffer to generate the checksum
+	// that we'll tack onto the end of the ticket.
+	checksum := sha256.Sum256(checksumBuf.Bytes())
+
+	// Now that we have the checksum, we'll write the raw ticket and the
+	// checksum to the buffer that we'll use for encoding.
+	//
+	// Similar to Bitcoin, we'll add a version so the prefix of the string
+	// is always the same.
+	if _, err := encodeBuf.Write(encodingVersion); err != nil {
+		return "", err
+	}
+	if err := SerializeTicket(&encodeBuf, t); err != nil {
+		return "", err
+	}
+	if _, err := encodeBuf.Write(checksum[:checksumLen]); err != nil {
```

Comment:
Looks like this could all be replaced with `base58.CheckEncode`?

Reply:
Can't use that here since we don't base58 encode the prefix itself. The prior version of this selected characters such that their encoding using base68 would _always_ be "sidecar". In this case, we just append the prefix to the encoding, then include it in the checksum.